### PR TITLE
Remove DeviceProxy pointer in ClusterBase

### LINKED ISSUE #19259: Make `OperationalDeviceProxy` shareable by multiple applications, and delivers its callbacks to multiple applications
#### Problem
* 1: ~~If an application context is released before the pairing completes, will render the pointer passed into `OnDeviceConnected` dangling~~
* 2: in `OperationalDeviceProxy::DequeueConnectionCallbacks` we may use member fields which is already destructed
  * `cb->mCall(cb->mContext, mPeerId, error)`
    * We are loop for all callbacks here, but `this` object may already destructed by previous callback, so `mPeerId` may not valid here
  * `cb->mCall(cb->mContext, this)`
    * Obviously, `this` may not valid here

 #### Proposed Solution
* ~~Implement a type-safe and RAII style callback queue using `IntrusiveList` to solve problem 1~~ 
* When implementing `DequeueConnectionCallbacks`, split the function into 2 parts:
  1. Move the callback queue to a stack variable
  2. Use an static member function to call all callback, to ensure that no member fields is used

### PR DESCRIPTION
**Pull request by kghost**
#### Problem
Pave the path to fix #19259

#### Change overview
Remove `DeviceProxy*`  in `ClusterBase` class, use a session handle ref instead.

~~@andy31415 Not sure how to generate files under `scripts/idl/tests/outputs`~~
@bzbarsky-apple Need to fix `src/darwin/Framework/CHIP/CHIPCluster.mm` but I don't understand object-c at all.

#### Testing
Refactor, can be verified by unit tests.

### CODE REVIEW

File: examples/chef/.shell_env [hunk 1]
```diff
@@ -0,0 +1,51 @@
+SHELL=/usr/bin/zsh
```
Comment (andy31415) [inline examples/chef/.shell_env:1]: this and other `chef/.*` files should not be checked in. I believe they should not be created anymore by latest chef builds.

**Review by mrjerryjohns — APPROVED** [review]
LGTM, but would like @bzbarsky-apple to review as well.

**Review by bzbarsky-apple — COMMENTED** [review]
Happy to do the Darwin bits of this once we sort out how the cluster bits are initialized; the exact Darwin changes will depend very heavily on that.

File: src/app/zap-templates/templates/app/CHIPClusters.zapt [hunk 2]
```diff
@@ -18,7 +18,8 @@ namespace Controller {
 class DLL_EXPORT {{asUpperCamelCase name}}Cluster : public ClusterBase
 {
 public:
-    {{asUpperCamelCase name}}Cluster() : ClusterBase(app::Clusters::{{asUpperCamelCase name}}::Id) {}
+    {{asUpperCamelCase name}}Cluster(DeviceProxy * device, EndpointId endpoint) : ClusterBase(*device->GetExchangeManager(), device->GetSecureSession().Value(), app::Clusters::{{asUpperCamelCase name}}::Id, endpoint) {}
```
[the remark this answers is not in the document]
Comment (kghost) [reply src/app/zap-templates/templates/app/CHIPClusters.zapt:21]: The goal is remove `device`, and use `ExchangeManager &, const SessionHandle &` instead

[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/zap-templates/templates/app/CHIPClusters.zapt:21]: OK, so why not do that?  Why are we doing this intermediate DeviceProxy * step?

[the remark this answers is not in the document]
Comment (kghost) [reply src/app/zap-templates/templates/app/CHIPClusters.zapt:21]: Will do it

File: scripts/idl/tests/outputs/several_clusters/jni/FirstClient-InvokeSubscribeImpl.cpp [hunk 3]
```diff
@@ -27,9 +27,8 @@ using namespace chip::Controller;
 JNI_METHOD(jlong, FirstCluster, initWithDevice)(JNIEnv * env, jobject self, jlong devicePtr, jint endpointId)
 {
     chip::DeviceLayer::StackLock lock;
-    FirstCluster * cppCluster = new FirstCluster();
-
-    cppCluster->Associate(reinterpret_cast<DeviceProxy *>(devicePtr), endpointId);
+    DeviceProxy * device = reinterpret_cast<DeviceProxy *>(devicePtr);
+    FirstCluster * cppCluster = new FirstCluster(*device->GetExchangeManager(), device->GetSecureSession().Value(), endpointId);
```
Comment (bzbarsky-apple) [inline scripts/idl/tests/outputs/several_clusters/jni/FirstClient-InvokeSubscribeImpl.cpp:31]: Can we get these generated files added to the generated file list?

File: src/controller/CHIPCluster.h [hunk 4]
```diff
@@ -397,17 +381,27 @@ class DLL_EXPORT ClusterBase
...
+                EndpointId endpoint) :
+        mExchangeManager(exchangeManager),
+        mSession(session), mClusterId(cluster), mEndpoint(endpoint)
+    {}
+
+    Messaging::ExchangeManager & mExchangeManager;
+
+    // Since cluster object is ephemeral, the session shall be valid during the entire lifespan, so we do not need to check the
+    // session existence when using it. For java and objective-c binding, the cluster object is allocated in the heap, such that we
+    // can't use SessionHandle here, in such case, the cluster object must be freed when the session is released.
```
Comment (bzbarsky-apple) [inline src/controller/CHIPCluster.h:401]: Follow-up: How would a consumer know when the session is released?

**Review by bzbarsky-apple — APPROVED** [review]

File: scripts/idl/tests/outputs/several_clusters/jni/FirstClient-InvokeSubscribeImpl.cpp (hunk 3 above)
Comment (kghost) [reply scripts/idl/tests/outputs/several_clusters/jni/FirstClient-InvokeSubscribeImpl.cpp:31]: Not sure about how to do this, @andy31415 would you please create a following up PR for it.

**Comment by woody-apple** [comment]
Needs to go in SVE, given high likelihood of divergence.

**Pull request merged**